File: server/Services/KeepService.cs

namespace keepr.Services;

public class KeepService
{
    private readonly KeepRepository _repository;

    public KeepService(KeepRepository repository)
    {
        _repository = repository;
    }

    internal List<Keep> GetKeeps()
    {
        List<Keep> keeps = _repository.GetKeeps();
        return keeps;
    }

    internal Keep GetKeepById(int keepId)
    {
        Keep keep = _repository.GetKeepById(keepId);
        if (keep == null)
        {
            throw new Exception("Unable to find keep!");
        }
        return keep;
    }

    internal Keep CreateKeep(Keep keepData, string userId)
    {
        keepData.CreatorId = userId;
        Keep keep = _repository.CreateKeep(keepData);
        return keep;
    }

    internal Keep UpdateKeep(Keep keepdata, int keepId, string userId)
    {
        Keep keepToUpdate = GetKeepById(keepId);
        if (keepToUpdate.CreatorId != userId)
        {
            throw new Exception("You cannot update what is not yours!");
        }
        keepToUpdate.Id = keepId;
        keepToUpdate.Name = keepdata.Name ?? keepToUpdate.Name;
        keepToUpdate.Description = keepdata.Description ?? keepToUpdate.Description;
        keepToUpdate.Img = keepdata.Img ?? keepToUpdate.Img;

        Keep keep = _repository.UpdateKeep(keepToUpdate);
        return keep;
    }

    internal string TrashKeep(int keepId, string userId)
    {
        Keep keep = GetKeepById(keepId);
        if (keep.CreatorId != userId)
        {
            throw new Exception("You cannot delete what is not yours!");
        }
        _repository.TrashKeep(keepId);
        return "Deleted Keep";
    }

    internal List<KeptVaultKeep> GetKeepsByVaultId(int keepId)
    {
        List<KeptVaultKeep> keeps = _repository.GetKeepsByVaultId(keepId);
        if (keeps == null)
        {
            throw new Exception("Unable to find keeps!");
        }
        return keeps;
    }

    internal Keep GetKeepByVaultKeep(int keepId)
    {
        Keep keep = _repository.GetKeepById(keepId);
        return keep;
    }

    internal List<Keep> GetKeepsByProfile(string profileId)
    {
        List<Keep> keeps = _repository.GetKeepsByProfile(profileId);
        return keeps;
    }

    internal List<Keep> GetMyKeeps(string userId)
    {
        List<Keep> keeps = _repository.GetAccountKeeps(userId);
        return keeps;
    }


    internal Keep IncrementViews(int keepId)
    {
        _repository.IncrementViews(keepId);

        Keep keep = GetKeepById(keepId);

        return keep;
    }
}

File: client/src/services/KeepService.js
import { AppState } from "../AppState.js"
import { Keep } from "../models/Keep.js"
import { logger } from "../utils/Logger.js"
import { api } from "./AxiosService.js"

class KeepService{
    unsetActiveKeep() {
        AppState.activeKeep = null
    }
    setActiveKeep(keepId) {
        AppState.activeKeep = null
        const keep = this.getKeepById(keepId)
        const activeKeep = new Keep(keep)
        AppState.activeKeep = activeKeep
    }

    async getKeepsByVault(vaultId) {
        const response = await api.get(`api/vaults/${vaultId}/keeps`)
        const keeps = response.data.map(keep => new Keep(keep))
        logger.log("keeps", keeps)
        AppState.keepsInVault = keeps
    }

    async getKeeps(){
        const response = await api.get("api/keeps")
        const keeps = response.data.map(keep => new Keep(keep))
        AppState.keeps = keeps
    }

    async updateKeep(keepData, keepId){
        const userId = AppState.account.id
        this.getKeepById(keepId)
        const keepcheck = AppState.activeKeep
        if(userId != keepcheck.creatorId){
            throw new Error("You can't edit what isn't yours!")
        }
        const response = await api.put(`api/keeps/${keepId}`, keepData)
        const keep = new Keep(response.data)
        const updated = AppState.keeps.findIndex(keep => keep.id = keepId)
        AppState.keeps.splice(updated, 1)
        AppState.keeps.push(keep)
    }
    async getKeepById(keepId) {
        const response = await api.get(`api/keeps/${keepId}`)
        const keep = new Keep(response.data)
        AppState.activeKeep = keep
    }

    async createKeep(keepData){
        const response = await api.post("api/keeps", keepData)
        const keep = new Keep(response.data)
        AppState.keeps.push(keep)
    }

    async trashKeep(keepId){
        const userId = AppState.account.id
        this.getKeepById()
        const keepCreator = AppState.activeKeep.creatorId
        if(userId != keepCreator){
            throw new Error("You can't delete what isn't yours!")
        }
        await api.delete(`api/keeps/${keepId}`)
        const trash = AppState.keeps.findIndex(keep => keep.id = keepId)
        AppState.keeps.splice(trash, 1)
        AppState.activeKeep = null
    }
}

export const keepService = new KeepService()